### ISSUE TITLE
Unit testing tool for SMVC

### ISSUE THREAD

**Issue by nekomajin** [nova-framework/framework#216]
Hi Guys. For those who don't know yet: SMVC has from today a Gitter chatroom (#214) [![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/simple-mvc-framework/framework?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge).

One issue was whether we want to use for the SMVC a Unit Testing tool ([Link to Gitter Chat](https://gitter.im/simple-mvc-framework/framework?at=5650a382541ec57a66e2eda8)).
I am definitely sure: It improves the code (style) and unifies it. Moreover, the question was how we strike a balance between PHP 5.5, 5.6 and 7. Benefits us a CI tool? And if so, which one? My suggestions would be [Travis CI](https://travis-ci.org/) and [Jenkins](https://jenkins-ci.org/). Both work very well together with GitHub.

What do you think about it?

**Comment by volter9** [comment]
+1 for CI and unit testing.

**Comment by tomvlk** [comment]
+1 for CI unit testing. Maybe also code quality checks?

**Comment by volter9** [comment]
I think, code quality checks are BS. Every person has its own preferences on what's high quality and what's not. Therefore, maybe I don't know a lot about it. How does these tools determine code quality?

P.S. Some of these tools have warning noted like: "this analysis doesn't replace code reviews"

**Comment by geomorillo** [comment]
I really have to to agree with volter9, really good code comes with experience, however the tools helps to identify problems and areas where you can optimize.

**Comment by nekomajin** [comment]
Of course you are right. Code quality also comes from experience to write the code. I meant that we should test the code. And thus improve the quality.
By the way: has anyone concrete experience with PHP unit tests?

**Comment by tomvlk** [comment]
Of course code quality test is not fully correct etc. But could still be usefull.
Anyway, I did a bit with PHPUnit, but not that much..

**Comment by volter9** [comment]
I have intermediate amount of experience with TDD and PHPUnit unit testing. I think unit testing would make SMVCF modular and improve overall quality, but will increase the complexity and code base, if done in the right way.

**Comment by volter9** [comment]
There's lots of places in SMVCF where it would be really hard to unit test. Good examples are: Config, View, Router, and others.

Most of these components are very tightly coupled and dependent on something. Let's take Config for example. Config contains only one method that creates lots of global constants. Yeah, you could check presence of all these globals, but only once. If there's going to be other classes dependent on Config that are going to be ran before Config you'll get error.

Let's take View. View is depending Config. Let's pretend that Config is already loaded and we're going to test View. It's dependent on file system, hardcoded path, and constants defined globally by Config.

Basically my point is that it would be hell of a job.

**Comment by nekomajin** [comment]
@volter9 You're absolutely right. It is a tough job. But "even the longest journey begins with the first step". We can start with small things and increase slowly. Like @daveismyname's work on the Framework: Steadily becoming better.
And because of this thing with the file system paths: I have read about [mock objects (Wikipedia)](https://en.wikipedia.org/wiki/Mock_object). Perhaps we can use it?

**Comment by daveismyname** [comment]
I've never used unit testing or done TDD but I completely agree this should be the direction we head down, I need to spend some time and get upto speed with that so I can understand what's involved and the limitations of the current setup to that end.

If it turns out we need to make core changes to the framework then so be it, if that's the case it would make sense to be a new release.

**Comment by jimgwhit** [comment]
This makes absolutely no sense, take the PSR -2 spec for example, it is a suggestion only to indent a certain way and that sort of thing. I can write my code perfect to the PSR spec, or have  a page of code that looks like a bowl of spaghetti, and either will still work as long as the code is correct. I declare some of these things modern people come up with!

**Comment by geomorillo** [comment]
So, you like it or not? :)

**Comment by jimgwhit** [comment]
From doing some googling it looks like unit testing can be completely separate from the framework, if someone wanted that they could do an addin themselves. Furthermore!, 2 more exclamation points, I fail to see what the big deal is with unit testing, I use NetBeans as an IDE and it catches mistakes as I go people please! People people learn how to code correctly and stop worrying about stupid unit testing GEES. This almost sounds as bad has some of the stuff on the laravel forum.  Why don't you fork the framework, and do whatever it is you want to do leave the current framework alone. Write your own framework based on this whatever. Besides there is no common sense being used
whatsoever, I could see unit testing Zend, but SMVC is a very small framework how in the heck can you goof up anything. If you can't write a small application with this framework you do not need to be programming rather you need to get another job doing completely something different.
But some people don't think things through.

**Comment by jimgwhit** [comment]
Are you going to unit test a raw PDO single page application good grief people. You forget the framework is very basic in nature and it's based on PDO. What is there to unit test????!!!!!

**Comment by volter9** [comment]
@jimgwhit I'm kind of agree with you about unit testing (we love SMVCF for simplicity), but I think you got unit testing wrong.

> unit testing can be completely separate from the framework

It's pointless to write unit tests separate from framework. After every change you'll have to correct unit tests. That's would be very hard.

> I use NetBeans as an IDE and it catches mistakes as I go people please!

Unit testing isn't about syntax errors, it's about logical errors and testing small units to work as you expect them to work.

> Are you going to unit test a raw PDO single page application good grief people. You forget the framework is very basic in nature and it's based on PDO. What is there to unit test????!!!!!

I think OP suggests unit testing for **framework's core** not for the application domain code.

**Comment by jimgwhit** [comment]
Unit testing has to do with application code you're running a very big app you unit test chunks of it as you go.  Why would the core need unit testing? If Dave wanted to test the framework he could and then simply let us know what files had to change. 
I mess around with asp.net sometimes, and it has debug mode. Well doesn't PHP have errors, and error catching? To me that is exactly the same as testing, no errors equals Pass, errors equals fail. Just curious why does some people try to make things harder and more complicated than they need to be? And whatever happen to if it's working it's okay? One of my production sites has been going 3 years with no errors, no bad pages, on nothing out of the ordinary. I would say it does not need no unit test.

**Comment by bvqbao** [comment]
But with unit test we can make testing automatic, right? Dave doesn't need
to tell us which files had been changed. He just launch the tests with a
single command.

One of the points of unit test is to make sure that all of the correctly
executed scenarios get executed correctly after the changes someone had
made. With manual testing, we need to do it by hand with every changes
made. It's painful.

I agree that unit test is overkill for small projects. But if someone who
loves SMVC want to build a rather big website on it, then it's would be
great to give him the ability to do unit testing with SMVC.

**Comment by daveismyname** [comment]
personally I think it makes a lot of sense to have automated testing, it's all to easy to miss something when testing manually.

I don't want to over complicate the framework either but if it can be used without completly changing the framework then that's got to be a good thing.

The framework as it's released it quite simple I think we all agree on that but your end product may not be so simple.

**Comment by jimgwhit** [comment]
You are referring to automated testing but by choice correct, like there would be a button to press to run tests or something. Even laravel has test that people can run but it is completely by choice. Have a look at the laravel  Docs and get an idea, but my no means do you have to run a test it is all up to the development. 
There should be a way to add in some testing without modifying the current framework.

**Comment by daveismyname** [comment]
no one is suggesting it's forced, it would be optional as I understand it in the current format the framework is not testable that is what we are suggesting adapting so it becomes testable. At no point will it be forced.

**Comment by jimgwhit** [comment]
Also could unit testing possibly be an addin only, someone could include it in composer duing an install? You mentioned having an alternative place  for addin packages so the original framework would not get bloated.

**Comment by daveismyname** [comment]
that section is live and lives here http://simplemvcframework.com/addons

the phpunit could be yes, but as I understand it in the current format the framework is not testable

**Comment by jimgwhit** [comment]
And weather I use it or not please do a video on usage.

**Comment by daveismyname** [comment]
Yes that would be useful for everyone including me lol

**Comment by tomvlk** [comment]
You can add the PhpUnit to the Dev dependencies in composer file. This will only install the PhpUnit needed files when installing with dev dependencies. 
I think you can test a part of the framework right now, you can simulate a route call, and check if routing will work correctly, and if the url class contains the right contents after routing has been done.

**Comment by jimgwhit** [comment]
At least https://phpunit.de/getting-started.html makes more sense than the original poster. But somehow it should still be an optional install.

**Comment by rocketpastsix** [comment]
I made a fork of 2.2 and will start working on integrating the core PHPunit files in with SMVC

**Comment by nekomajin** [comment]
Sorry for the late reply. I had the last few days a lot to do (training, more on that later). I have nevertheless observed how the debate has evolved and want to express myself as a contribution creator also. For clarity, I will divide my answer into several areas.

##### Thanks to the Community

First I want to thank the community that my proposal was received positively by the most people. The majority has understood what I want to achieve with this proposal.

##### What exactly is the problem?

Next, I turn to @jimgwhit: What is your problem? I bring a proposal for discussion, have neither push a PR nor anything changed on framework itself. Nevertheless, I will verbally attacked by you. Why are you doing such a thing?
This makes me sad. I want to help the community, contribute to something big and you punch me down! Do you find that right?

##### Quotations and my opinion on it

> People people learn how to code correctly and stop worrying about stupid unit testing GEES.

Do you know what you're talking about? Have you dealt with the subject or really just

> From doing some googling [...]

If so I would advise you the next time more closely to Google ([How to search on Google](https://support.google.com/websearch/answer/134479?hl=en)) and the way to read references more closely looking at that ([unit testing, wikipedia](https://en.wikipedia.org/wiki/Unit_testing)). There you will also find [additional literature to unit testing](https://en.wikipedia.org/wiki/Unit_testing#Notes).

##### Opinions of people you do not know

> If you can't write a small application with this framework you do not need to be programming rather you need to get another job doing completely something different.

For your info: I'm a trained computer scientist with a focus on PHP (that's why I did not have time to respond, as mentioned above). In the training I haven't learned how I program the next Windows or Linux, develop **the ultimate tool** to combat [HIV](https://en.wikipedia.org/wiki/HIV) or publish help against the war and poverty in our world.

But a focus of my project work for completion of the training was the subject *unit tests*. And so I came up with the idea that propose here. We are a great community and help each other. The motto of GitHub is *Build software better, together.* And nothing else I want.

> You have to understand something some people hear a latest fad going on and have to jump all over it. [...]  So basically some preppy heard about unit testing and decided that has to be the latest thing. That does not make it right.

And again, just for your info: Unit testing is not **the latest crazy** :poop:. It's a chance to test your code to find errors (which are also found in PHP) and to improve the code and gain experience.

[Sebastian Bergman](https://sebastian-bergmann.de/) published [PHPUnit](https://phpunit.de/) the first time [in 2004](https://sebastian-bergmann.de/archives/300-PHPUnit-1.0.0-Released.html). Is that your definition of **the latest crazy** :poop:?

##### Requirements for the Community

How often do you require

> Please make sure it works on PHP 7 as well Thanks.

We do it. Damn one. We do this always! And guess what: With [TravisCI](https://travis-ci.org/) much work is taken off us. We no longer need to set up multiple virtual boxes and test each new PHP 7 RC. [That makes TravisCI for us!](https://docs.travis-ci.com/user/languages/php) Surprised?

##### I have finished! ([Trapattoni, YouTube](https://www.youtube.com/watch?v=Bqp64q7kHmw))

Finally, one more thing: I know you don't like laravel. That's what you now have sufficiently announced frequently. However, this framework has a right to exist. Let them make their things. Just as we do ours.

> And weather I use it or not please do a video on usage.

Seriously? :see_no_evil:

**Comment by rocketpastsix** [comment]
To follow @nekomajin's post, I was literally on a meeting with Sebastian Bergmann this morning. PHPUnit is far from dead, or going away, and in fact, on PHP 7 it is sped up. So its definitely a tool we should look to integrate into the framework. 

I also believe that that we should look at codeception, travis, or another CI type server for automation testing as well.

**Comment by jimgwhit** [comment]
@rocketpastsix You do realize you can do that to your own project with composer, right, why bloat the framework with stuff that some people don't want?  We may as well change the name to laravel junior or zend junior.  
I chose SMVC because it's small and not bloated.  
@daveismyname can you do two frameworks, one bloated for people who want SMVC to turn into laravel, and keep this one the way it is? Just call the other one simple-mvc-framework-bloated-to-laravel-size-for-@rocketpastsix

**Comment by daveismyname** [comment]
@jimgwhit come on you're just being silly now. Yes that can be dependencies but the framework still needs to be testable.

I really don't understand your resistance, this is not a fad or some random though it's makes a lot of sense, it's not going to bloat the framework by changing classes so they become testable once that's done the end user can choose what testing kit they use if any. As I've said multiple times this will not be forced! don't want to use unit testing or anything like that then don't.

I really wish you would not so rude with people, it's really not called for. We are a community after all everyone has their own opinions, they may be different to yours and that should be OK.

**Comment by nhymxu** [comment]
I don't thing unit testing is bloat?
It's tool for developer keep control function work well. You can easy delete unit tests folder on production stage.

**Comment by jimgwhit** [comment]
After a re-read, darn that was pretty silly. But I bet it made some folks laugh.

**Comment by nekomajin** [comment]
@LuckyCyborg How do you feel about the subject unit tests? Surely you use this during development. May also in conjunction with [Travis CI](https://travis-ci.org/) or another tool?

I ask because, thanks to your help (and of course @tomvlk), it has been a lot of changes. Meanwhile, PHP 7.0.1 is [released](http://php.net/archive/2015.php#id2015-12-17-1). And so that @jimgwhit can continue to work in peace :yum: and we should simplify the testing work, I suggested some time ago that we use a tool like Travis CI.

I want these tests for both SMVC (or perhaps soon [Nova](http://www.poll-maker.com/results518091x6D6168D4-22#tab-2) Framework :wink:) 2.2 and 3.0.

**Comment by tomvlk** [comment]
@nekomajin I think unit tests are really nice to have, in combination with travis. We could create tests to simulate a request in the router for example. Not that hard to do I think. I can't check my code constantly on PHP 7 compatibility, but travis can without doing any installations at all!

**Comment by LuckyCyborg** [comment]
Hello to all,

sure that i used Unit Tests in the Projects I worked. Travis CI and other things, even in-house ones. 

Yet, I do not feel the lack of Unit Testing in working on SMVC3, because the Project is so small, right now.

But I feel the lack of a Debugger or Profiler, even a small one, from the day one. 

In other hand, the Unit Tests are a good thing to test the overall design of a Project, that it reach the goals.

Yet, no automated tests can replace the the programmers experience and its skills. I seen hundred styles of writing code, and some being literally, from the day one, spaghetti code, even they pass the automated tests.

At the SVMC dimension, a experimented programmer can understand well, all its design elements in several day, max.

What SVMC lack right now, in my opinion, as a programmer POV, is to have incorporated a little debugger/profiler; then you can think to Unit Tests and other nice things.

And someone to implement **Database Migrations**, for example using **Phinx** https://phinx.org/

Finally, to know that the companies use to have two software teams: one of **programmers**, one of **testers**, yet I don't seen an automated Unit Tests engine being able to replace an experienced Tester.

All the best!

**Comment by geomorillo** [comment]
what do you mean by Database Migrations? how does it should be implemented?

**Comment by LuckyCyborg** [comment]
@geomorillo A thing very useful in development, when you work under a framework and extremelly useful when you work with a Modular Framework, like SVMC will be.

For example, let's say that I have a website, where is is installed a module **Dogs v15.4** developed by our friend Jim. Then,  e publish **Dogs v16.0** with loads of enhancements and optimizations. 

Me, I want to update my module **Dogs** to the superior version, but there are some small changes in Dogs tables, in database. I can modify the database manually, or I can use Database Migrations, to automate the upgrade, and it to make the required modifications. 

You can also create database tables, modify them, delete them, etc.

That's do, in few words, that Database Migrations.

**Comment by geomorillo** [comment]
But shouldn't the programmer do this manually i mean implement the migration by code or did you mean to do this migration automatically like the framework detecting automatically everithing?

**Comment by LuckyCyborg** [comment]
@geomorillo No, it is not an automatically detection. The programmer write a series of Migrations, practically, a series of classes, and the **Phinx**, for example, know ho to execute them in a specific order, at command.

For example, you seen ever Drupal how update its database after upgrading some modules?
In fact, what execute it, when do update to database, is it called Database Migrations.

**Comment by geomorillo** [comment]
But you said : "What SVMC lack right now, in my opinion, as a programmer POV, is to have incorporated a little debugger/profiler; then you can think to Unit Tests and other nice things.

And someone to implement Database Migrations, for example using Phinx https://phinx.org/"

I just thought you mean do this automatically, i was wrong

**Comment by geomorillo** [comment]
But maybe using Entity, doing it automatically could be done.

namespace App\Models\Entities;

use Smvc\Database\Entity;

class Car extends Entity
{
    public $carid;
    public $makeid;
    public $model;
    public $type;
    public $costs;
}

**Comment by LuckyCyborg** [comment]
Nope, the Migrations are another things. :smirk: 

Go to Phinx website and see how the things works. They have a good documentation and video tutorials.

BTW, even latest CakePHP series, **3.0**, use the Phinx at base of their Migrations.

**Comment by geomorillo** [comment]
Actually after reading a bit.... the ones you call database migration are actually code based migrations, and the ones i'm talking about are automatic migrations, they are not the same but  do the same thing.

**Comment by LuckyCyborg** [comment]
@geomorillo You, really, but really really, don't want to leave a Framework to automatically calculate tables structure modifications and to modify you database tables structure, trust me! :smirk:

**Comment by geomorillo** [comment]
lol

**Comment by tomvlk** [comment]
I think we can still think about the unit testing, at least for the database engine's it could be usefull, to keep testing the engine's with every code change. Of course it doesn't replace the normal testing, but it will consistant the testing, also after some time, or when new PHP is released.

**Comment by nekomajin** [comment]
@tomvlk I agree with you. For every new class / method a unit test should be written, in order to keep the consistency.

**Comment by tomvlk** [comment]
@nekomajin I'm already testing some stuff. But is a bit complex. I will try to make the sqlite service soon too.

**Comment by nekomajin** [comment]
I'm working on small, simple unit tests in conjunction with Travis CI. Based on this, I will develop further testing. First PRs in my forked repository after Christmas.

**Comment by LuckyCyborg** [comment]
@nekomajin And what you want to test on SMVC3/Nova, my German friend?

Routing? Controllers? Views rendering? Events? That's some of the things usually almost impossible to test with automated methods, because of dynamically and distributed response.

Oh, we have also the **DBAL**. What is to test in DBAL, when all of its code can be read in five minutes and analyzed in twenty?

**Comment by geomorillo** [comment]
Using a cannon to kill a bug lol

**Comment by tomvlk** [comment]
@nekomajin + all: https://github.com/simple-mvc-framework/framework/pull/308

**Comment by nekomajin** [comment]
@LuckyCyborg There are people who are born and master PHP. I don't belong. By SMVC / Nova I learn very much about "how to use a framework" and PHP. And by writing tests, I also check whether I understood what makes a method.

@tomvlk Very cool and many thanks :thumbsup:
